### PR TITLE
GUVNOR-3090: Guided Decision Table: Problems deleting rows from decis…

### PR DESCRIPTION
**Pull request by markcoble**
…ion table

### CODE REVIEW

**Comment by markcoble** [comment]
@jomarko I am starting on rhbrms-2909 today.  Just can't change the status to inprogress for some reason.

**Comment by jomarko** [comment]
@markcoble I think there are missing acks. I will have a look on this PR. Please feel free to mark me also in the one for  RHBRMS-2909. Thanks for solving these issues.

**Comment by jomarko** [comment]
@markcoble Hi, please could you provide more details how to reproduce the issue? I had look on GUVNOR-3090 but I am not able to reproduce the original bug, or better to say, I am not sure what was the original bug. With 7.3.0-SNAPSHOT, I opened GDT with 10 rows, I was deleting one after one but I haven't noticed any error, that is why I am asking for more details about the issue.

**Comment by manstis** [comment]
@jomarko if, like me, you selected a row, deleted it; selected a row, deleted it etc you will not find any problem. The problem is select a row, delete it from the edit menu; then (without selecting anything) delete again from the edit menu. The cause was that when a row/column/cell was deleted and the selections updated (i.e. remove a selection for a deleted row/column/cell) the menus were not being refreshed.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 1]
```diff
@@ -1641,4 +1655,17 @@ public void testRefreshView() {
         verify(parent).refreshScrollPosition();
         verify(layer).draw();
     }
+    
+    private void checkDTEventFired(int times){
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: There are lots of events relating to the Decision Table; I'd prefer this to be renamed something more explicit like ```checkDecisionTableSelectionsChangedEventFired(int times)```.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 2]
```diff
@@ -1641,4 +1655,17 @@ public void testRefreshView() {
...
+        final ArgumentCaptor<DecisionTableSelectionsChangedEvent> dtSelectionsChangeEventCaptor = ArgumentCaptor.forClass(DecisionTableSelectionsChangedEvent.class);
+ 
+        verify(decisionTableSelectionsChangedEvent,times(times)).fire(dtSelectionsChangeEventCaptor.capture());
+        
+        final DecisionTableSelectionsChangedEvent dtSelectionChangeEvent = dtSelectionsChangeEventCaptor.getValue();
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: If the ```times``` is more than one there will be multiple values; therefore it's probably best to use ```dtSelectionsChangeEventCaptor.getAllValues()```, check the returned ```List``` size equals ```times``` and that each instance's ```presenter``` equals ```dtPresenter```

**Comment by jomarko** [comment]
@manstis sorry, just to be sure I recorded video of 7.3.0-SNAPSHOT https://webmshare.com/JGqMW The only problem I can see is that the `Edit Menu - Delete Cells` is not disabled after one row was deleted, there is no new selection after deleting row. Are we fixing  disabling/enabling `Edit Menu - Delete Cells` menu item appropriately after deleting rows? The same for columns, cells?

**Comment by manstis** [comment]
@jomarko Hi.. so after you've deleted the first row there are no rows selected; yet the "delete row" entry in the menu is still enabled. What row should be deleted!?! The issue is that after you've deleted a row the menu should be disabled. The JIRA's description that you cannot delete rows is indeed a symptom of the bug and not the bug itself. In your video you deleted a couple of rows from the (not disabled) menu and then selected a row. If you'd carried on using the menu without selecting a row you'd have found the symptom "unable to delete rows"...

**Comment by jomarko** [comment]
@manstis thanks, I updated the jira to prevent misunderstanding in future.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java (hunk 1 above)
Comment (markcoble) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: @manstis i agree as well.  originally thought of using generic argument type to allow for the multiple types but then decided until needed for this JUNit I';d keep it simple.  Will rename.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java (hunk 2 above)
Comment (markcoble) [reply drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: @manstis thanks for clarifying that bit of logic.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java [hunk 3]
```diff
@@ -379,6 +379,28 @@ public boolean isRowDraggingEnabled() {
             public boolean isColumnDraggingEnabled() {
                 return access.isEditable();
             }
+            
+            @Override
+            public Range deleteCell(int rowIndex,
+                                    int columnIndex) {
+
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java]: Blank line

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java [hunk 4]
```diff
@@ -379,6 +379,28 @@ public boolean isRowDraggingEnabled() {
             public boolean isColumnDraggingEnabled() {
                 return access.isEditable();
             }
+            
+            @Override
+            public Range deleteCell(int rowIndex,
+                                    int columnIndex) {
+                
+            	Range cellRange = super.deleteCell(rowIndex, columnIndex);
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java]: Parameters split to new lines on "," and aligned.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 5]
```diff
@@ -1641,4 +1655,17 @@ public void testRefreshView() {
         verify(parent).refreshScrollPosition();
         verify(layer).draw();
     }
+    
+    private void checkDTEventFired(int times){
+        final ArgumentCaptor<DecisionTableSelectionsChangedEvent> dtSelectionsChangeEventCaptor = ArgumentCaptor.forClass(DecisionTableSelectionsChangedEvent.class);
+ 
+        verify(decisionTableSelectionsChangedEvent,times(times)).fire(dtSelectionsChangeEventCaptor.capture());
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: Parameters split to new lines on "," and aligned.

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java [hunk 6]
```diff
@@ -341,13 +341,15 @@ void initialiseLockManager() {
                                         parent.getView().asWidget(),
                                         placeRequest,
                                         () -> currentPath.getFileName() + " - " + resourceType.getDescription(),
-                                        () -> {/*nothing*/}),
+                                        () -> {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java]: Any chance you could move this back onto a single line?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java [hunk 7]
```diff
@@ -468,10 +492,12 @@ void initialiseEventHandlers() {
         view.registerNodeMouseDoubleClickHandler((event) -> {
             if (view.isNodeMouseEventOverCaption(event)) {
                 if (isGridPinned()) {
-                    exitPinnedMode(() -> {/*Nothing*/});
+                    exitPinnedMode(() -> {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java]: Any chance you could move this back onto a single line?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java [hunk 8]
```diff
@@ -468,10 +492,12 @@ void initialiseEventHandlers() {
...
+                        /*Nothing*/});
                 } else {
                     enterPinnedMode(view,
-                                    () -> {/*Nothing*/});
+                                    () -> {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenter.java]: Any chance you could move this back onto a single line?

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 9]
```diff
@@ -407,11 +410,14 @@ public void checkRegisteredNodeMouseDoubleClickHandlerNotOverHeader() {
     public void link() {
         final GuidedDecisionTableView.Presenter dtPresenter2 = mock(GuidedDecisionTableView.Presenter.class);
         final GuidedDecisionTableView.Presenter dtPresenter3 = mock(GuidedDecisionTableView.Presenter.class);
-        final Set<GuidedDecisionTableView.Presenter> dtPresenters = new HashSet<GuidedDecisionTableView.Presenter>() {{
-            add(dtPresenter);
-            add(dtPresenter2);
-            add(dtPresenter3);
-        }};
+        final Set<GuidedDecisionTableView.Presenter> dtPresenters = new HashSet<GuidedDecisionTableView.Presenter>() {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: Is this Eclipse's formatting of static initialiser blocks (in anonymous sub-classes!)

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 10]
```diff
@@ -620,11 +626,14 @@ private void linkTables() {
         final GuidedDecisionTableView dtView3 = mock(GuidedDecisionTableView.class);
         final GuidedDecisionTableView.Presenter dtPresenter2 = setupPresenter();
         final GuidedDecisionTableView.Presenter dtPresenter3 = setupPresenter();
-        final Set<GuidedDecisionTableView.Presenter> dtPresenters = new HashSet<GuidedDecisionTableView.Presenter>() {{
-            add(dtPresenter);
-            add(dtPresenter2);
-            add(dtPresenter3);
-        }};
+        final Set<GuidedDecisionTableView.Presenter> dtPresenters = new HashSet<GuidedDecisionTableView.Presenter>() {
+
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: Is this Eclipse's formatting of static initialiser blocks (in anonymous sub-classes!)

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 11]
```diff
@@ -1360,9 +1374,12 @@ public void onDeleteSelectedCellsWithSelection() {
 
     @Test
     public void onDeleteSelectedCellsWithSelectionWithBooleanColumn() {
-        final AttributeCol52 column = new AttributeCol52() {{
-            setAttribute(RuleAttributeWidget.ENABLED_ATTR);
-        }};
+        final AttributeCol52 column = new AttributeCol52() {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: Is this Eclipse's formatting of static initialiser blocks (in anonymous sub-classes!)

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 12]
```diff
@@ -1374,13 +1391,17 @@ public void onDeleteSelectedCellsWithSelectionWithBooleanColumn() {
...
                                    any(Integer.class));
+        checkDTSelectionsChangedEventFired(1);
     }
 
     @Test
     public void onDeleteSelectedCellsWithSelectionsWithBooleanColumn() {
-        final AttributeCol52 column = new AttributeCol52() {{
-            setAttribute(RuleAttributeWidget.ENABLED_ATTR);
-        }};
+        final AttributeCol52 column = new AttributeCol52() {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: Is this Eclipse's formatting of static initialiser blocks (in anonymous sub-classes!)

File: drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java [hunk 13]
```diff
@@ -1426,9 +1449,12 @@ public void onDeleteSelectedCellsWithoutSelections() {
 
     @Test
     public void onDeleteSelectedColumnsWithSelections() throws ModelSynchronizer.MoveColumnVetoException {
-        final AttributeCol52 column = new AttributeCol52() {{
-            setAttribute("attribute1");
-        }};
+        final AttributeCol52 column = new AttributeCol52() {
```
Comment (manstis) [inline drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/test/java/org/drools/workbench/screens/guided/dtable/client/widget/table/GuidedDecisionTablePresenterTest.java]: Is this Eclipse's formatting of static initialiser blocks (in anonymous sub-classes!)

**Comment by markcoble** [comment]
@manstis all these formatting changes came from the latest Eclipse formatter I have.  I'll double-check though as I thought it was supposed to be mostly in line with that of Intellij.

**Comment by jomarko** [comment]
@manstis @markcoble we are lost, formatting guru @hasys is on PTO.

**Comment by manstis** [comment]
@markcoble OK.. np. Approved and merged. 

Please ensure the JIRA is moved to "Resolved" (fix version 8.0.0.beta1).

**Comment by manstis** [comment]
@markcoble @jomarko Should this also be on 7.3.x?

**Comment by jomarko** [comment]
@manstis in my opinion yes, issue is still present on 7.3.x branch

**Comment by manstis** [comment]
@jomarko I agree! 

@markcoble Please be sure to submit a PR for 7.3.x

**Pull request merged**